*This notebook is a guided project from datquest course (data analyst path)*

# Profitanle App Profiles for the App Store and Google Play Markets


For this project, we'll pretend we're working as data analysts for a company that builds Android and iOS mobile apps. We make our apps available on Google Play and the App Store.

In [15]:
import csv
file_ios=open('AppleStore.csv')
file_android=open('googleplaystore.csv')
read_ios=csv.reader(file_ios)
read_android=csv.reader(file_android)
data_ios=list(read_ios)
data_android=list(read_android)


In [12]:
def explore_data(dataset, start, end, rows_and_columns=False):
    dataset_slice = dataset[start:end]    
    for row in dataset_slice:
        print(row)
        print('\n') # adds a new (empty) line after each row

    if rows_and_columns:
        print('Number of rows:', len(dataset))
        print('Number of columns:', len(dataset[0]))

In [16]:
print('ios data:')
explore_data(data_ios,0,5)
print('android data:')
explore_data(data_android,0,6)
print('ios data columns:')
print(data_ios[0])
print('number of ios columns')
print(len(data_ios[0]))
print('android data columns:')
print(data_android[0])
print('number of android columns')
print(len(data_android[0]))

ios data:
['id', 'track_name', 'size_bytes', 'currency', 'price', 'rating_count_tot', 'rating_count_ver', 'user_rating', 'user_rating_ver', 'ver', 'cont_rating', 'prime_genre', 'sup_devices.num', 'ipadSc_urls.num', 'lang.num', 'vpp_lic']


['284882215', 'Facebook', '389879808', 'USD', '0.0', '2974676', '212', '3.5', '3.5', '95.0', '4+', 'Social Networking', '37', '1', '29', '1']


['389801252', 'Instagram', '113954816', 'USD', '0.0', '2161558', '1289', '4.5', '4.0', '10.23', '12+', 'Photo & Video', '37', '0', '29', '1']


['529479190', 'Clash of Clans', '116476928', 'USD', '0.0', '2130805', '579', '4.5', '4.5', '9.24.12', '9+', 'Games', '38', '5', '18', '1']


['420009108', 'Temple Run', '65921024', 'USD', '0.0', '1724546', '3842', '4.5', '4.0', '1.6.2', '9+', 'Games', '40', '5', '1', '1']


android data:
['App', 'Category', 'Rating', 'Reviews', 'Size', 'Installs', 'Type', 'Price', 'Content Rating', 'Genres', 'Last Updated', 'Current Ver', 'Android Ver']


['Photo Editor & Candy Camera

In [32]:
print(data_android[1047][2])
    

NaN


# ps
the google play data set has some duplicate entries that we ll work on remove later, for now we are going to confirm that by printing some duplicate.

In [17]:
for row in data_android[1:]:
    name=row[0]
    if name=='Instagram':
        print(row)

['Instagram', 'SOCIAL', '4.5', '66577313', 'Varies with device', '1,000,000,000+', 'Free', '0', 'Teen', 'Social', 'July 31, 2018', 'Varies with device', 'Varies with device']
['Instagram', 'SOCIAL', '4.5', '66577446', 'Varies with device', '1,000,000,000+', 'Free', '0', 'Teen', 'Social', 'July 31, 2018', 'Varies with device', 'Varies with device']
['Instagram', 'SOCIAL', '4.5', '66577313', 'Varies with device', '1,000,000,000+', 'Free', '0', 'Teen', 'Social', 'July 31, 2018', 'Varies with device', 'Varies with device']
['Instagram', 'SOCIAL', '4.5', '66509917', 'Varies with device', '1,000,000,000+', 'Free', '0', 'Teen', 'Social', 'July 31, 2018', 'Varies with device', 'Varies with device']


Counting the number of duplicates: 

In [18]:
duplicates=[]
unique=[]
for row in data_android[1:]:
    name=row[0]
    if name in unique:
        duplicates.append(name)
    else:
        unique.append(name)
print('nb of duplicate apps:', len(duplicates))
print('nb of unique apps:', len(unique))

nb of duplicate apps: 1181
nb of unique apps: 9660


We certainly need to remove duplicates in our cleaning data process, but we won't do that randomly.
when we printed the instagram app duplicates we noticed that the only difference between them was on the third column wich refers to the reviews. So the data was collected at different times. The best thing to do is to keep the latest version of the row I.E: the one with highest reviews value.


In [19]:
reviews_max={}
for row in data_android[1:]:
    name=row[0]
    n_reviews=row[3]
    if (name in reviews_max) and (reviews_max[name]< n_reviews):
        reviews_max[name]= n_reviews
    elif name not in reviews_max:
        reviews_max[name]=n_reviews
print(len(reviews_max))        

9660


In [20]:
android_clean=[]
already_added=[]
for row in data_android[1:]:
    name=row[0]
    n_reviews=row[3]
    if (n_reviews==reviews_max[name]) and (name not in already_added):
        android_clean.append(row)
        already_added.append(name)
print(len(android_clean))        
        

9660


In [23]:
def english_name(string):
    i=0
    for r in string:
        if ord(r)>127:
            i+=1
            if i>3:
                return False
    return True


In [44]:
english_name('Docs To Go™ Free Office Suite')


True

# filtring non english app names


In [24]:
#for android data set 
english_android=[]
for row in android_clean:
    name=row[0]
    if english_name(name):
        english_android.append(row)
#for ios data set
english_ios=[]
for row in data_ios[1:]:
    name=row[0]
    if english_name(name):
        english_ios.append(row)
#exploring the remaining data 
#for android
print(len(english_android))
#for ios
print(len(english_ios))
        
        
    

9615
7197


In [25]:
#last thing in our cleaning process is to keep only free apps data
#for both ios and android dataset
#for android
clean_android_vf=[]
for row in english_android:
    price=row[6]
    if price=='Free':
        clean_android_vf.append(row)
#for ios 
clean_ios_vf=[]
for row in english_ios:
    price=float(row[4])
    if price==0.0:
        clean_ios_vf.append(row)
#checking the length
print('the length for android cleaned dataset', len(clean_android_vf))
print('the length for ios cleaned dataset', len(clean_ios_vf))

the length for android cleaned dataset 8861
the length for ios cleaned dataset 4056


Our end goal is to add an app on both google play and the app store, and we want it to be profitable.
so we are going to define the types of apps that works well for both markets.
we are going to build two frequency table for each dataset.
for the ios dataset: using the prime genre column
for the android dataset: using the genre and category columns

In [26]:
def freq_table(dataset, index):
    freq_dict={}
    for row in dataset:
        column=row[index]
        if column in freq_dict:
            freq_dict[column]+=1
        else:
            freq_dict[column]=1
    return freq_dict   
#display table function
#this function transforms the frequence table into a list 
#of tuples, then sorts the list in descending order
def display_table(dataset, index):
    table = freq_table(dataset, index)
    table_display = []
    for key in table:
        key_val_as_tuple = (table[key], key)
        table_display.append(key_val_as_tuple)

    table_sorted = sorted(table_display, reverse = True)
    for entry in table_sorted:
        print(entry[1], ':', entry[0])

In [29]:
print('for the Genres column: \n')
display_table(clean_android_vf,9)
print('\n')
print('\n')
print('for the category column: \n ')
display_table(clean_android_vf,1)
print('\n')
print('\n')
print('for the prime-genre column: \n')
display_table(clean_ios_vf,11)

for the Genres column: 

Tools : 748
Entertainment : 538
Education : 474
Business : 407
Productivity : 345
Lifestyle : 345
Finance : 328
Medical : 312
Sports : 307
Personalization : 294
Communication : 287
Action : 275
Health & Fitness : 273
Photography : 261
News & Magazines : 248
Social : 236
Travel & Local : 206
Shopping : 199
Books & Reference : 190
Simulation : 181
Dating : 165
Arcade : 164
Video Players & Editors : 157
Casual : 155
Maps & Navigation : 124
Food & Drink : 110
Puzzle : 100
Racing : 88
Role Playing : 83
Libraries & Demo : 83
Auto & Vehicles : 82
Strategy : 80
House & Home : 73
Weather : 71
Events : 63
Adventure : 60
Comics : 54
Beauty : 53
Art & Design : 53
Parenting : 44
Card : 39
Casino : 38
Trivia : 37
Educational;Education : 35
Educational : 33
Board : 33
Education;Education : 30
Word : 23
Casual;Pretend Play : 21
Music : 18
Puzzle;Brain Games : 16
Racing;Action & Adventure : 15
Entertainment;Music & Video : 15
Casual;Brain Games : 12
Casual;Action & Adventure : 

# analyze
from the frequency table listed above.
for the prime_genre column (the only  one from the ios dataset) we have:
_The most common genre is games with runner-up of 2257 apps
_As a general impression we can say that the most used apps are designed for entertainment: (games:2257, entertainment:334, photo & video:167, social networking:143..)
_We can recommend the gamming genre 

for the category and genre column (android dataset/ google play)
_most common:tools

** the frequency tables generated reveal the most frequent app genres and not what genres have the most users, for that we need to includes data from the number of downloads**


The frequency tables we analyzed on the previous screen showed us that the App Store is dominated by apps designed for fun, while Google Play shows a more balanced landscape of both practical and fun apps. Now, we'd like to get an idea about the kind of apps with the most users.

One way to find out what genres are the most popular (have the most users) is to calculate the average number of installs for each app genre. For the Google Play data set, we can find this information in the Installs column, but this information is missing for the App Store data set. As a workaround, we'll take the total number of user ratings as a proxy, which we can find in the rating_count_tot app.

In [49]:
freq_tab_genre_ios=freq_table(clean_ios_vf,11)
for genre in freq_tab_genre_ios:
    total=0
    len_genre=0
    for row in clean_ios_vf:
        genre_app=row[11]
        if genre_app==genre:
            user_ratting=float(row[6])
            total+=user_ratting
            len_genre+=1
    avrg=total/len_genre
    print(genre ,':',avrg)

Productivity : 225.3709677419355
Catalogs : 364.77777777777777
Finance : 298.98809523809524
Book : 106.68181818181819
Games : 609.4714222419141
Social Networking : 646.0
Education : 683.3712121212121
News : 62.3448275862069
Navigation : 228.15
Food & Drink : 457.4651162790698
Shopping : 714.8595041322315
Weather : 1624.4516129032259
Photo & Video : 417.47904191616766
Health & Fitness : 234.02631578947367
Lifestyle : 1007.5
Travel : 175.89285714285714
Sports : 177.25316455696202
Entertainment : 179.31736526946108
Music : 618.8059701492538
Utilities : 1618.2844036697247
Reference : 2471.9
Medical : 36.75
Business : 183.6


from the table above we can recomande the reference apps genre wich have a high ratting average of 2471.9 or the weather one wich have a 1624.45 average

In [50]:
display_table(clean_android_vf,5)

1,000,000+ : 1395
100,000+ : 1024
10,000,000+ : 932
10,000+ : 904
1,000+ : 744
100+ : 613
5,000,000+ : 606
500,000+ : 494
50,000+ : 423
5,000+ : 400
10+ : 314
500+ : 288
50,000,000+ : 203
100,000,000+ : 188
50+ : 170
5+ : 70
1+ : 45
500,000,000+ : 24
1,000,000,000+ : 20
0+ : 4


In [53]:
#later we are going to need a list of unique genre(category)
duplicates_category=[]
unique_category=[]
for row in clean_android_vf:
    category=row[1]
    if category in unique_category:
        duplicates_category.append(category)
    else:
        unique_category.append(category)
print('nb of duplicate category:', len(duplicates_category))
print('nb of unique category:', len(unique_category))

nb of duplicate category: 8828
nb of unique category: 33


now let s calculate the average number of installs per app genre for the Google Play data set

In [62]:
#generating a frequency table for the category column
cat_freq=freq_table(clean_android_vf,1)
for category in unique_category:
    total=0
    len_category=0
    for row in clean_android_vf:
        category_app=row[1]
        if category_app==category:
            n_installs = row[5]
            n_installs = n_installs.replace(',', '')
            n_installs = n_installs.replace('+', '')
            total += float(n_installs)
            len_category+=1
    avg_nb_installs=total/len_category
    print(category, ':', avg_nb_installs)

ART_AND_DESIGN : 1986335.0877192982
AUTO_AND_VEHICLES : 647317.8170731707
BEAUTY : 513151.88679245283
BOOKS_AND_REFERENCE : 8767811.894736841
BUSINESS : 1712290.1474201474
COMICS : 817657.2727272727
COMMUNICATION : 38456119.167247385
DATING : 854028.8303030303
EDUCATION : 1820673.076923077
ENTERTAINMENT : 11640705.88235294
EVENTS : 253542.22222222222
FINANCE : 1387692.475609756
FOOD_AND_DRINK : 1924897.7363636363
HEALTH_AND_FITNESS : 4188821.9853479853
HOUSE_AND_HOME : 1331540.5616438356
LIBRARIES_AND_DEMO : 638503.734939759
LIFESTYLE : 1437816.2687861272
GAME : 15560965.599534342
FAMILY : 3696479.242695289
MEDICAL : 120616.48717948717
SOCIAL : 23253652.127118643
SHOPPING : 7036877.311557789
PHOTOGRAPHY : 17805627.643678162
SPORTS : 3638640.1428571427
TRAVEL_AND_LOCAL : 13984077.710144928
TOOLS : 10682301.033377837
PERSONALIZATION : 5201482.6122448975
PRODUCTIVITY : 16787331.344927534
PARENTING : 542603.6206896552
WEATHER : 5074486.197183099
VIDEO_PLAYERS : 24727872.452830188
NEWS_AND_

from here we can recommend the video_players category or communication 